plugins/org.eclipse.xtext/src/org/eclipse/xtext/generator/URIBasedFileSystemAccess.xtend

```diff
@@ -73,9 +77,10 @@ class URIBasedFileSystemAccess extends AbstractFileSystemAccess2 {
 	
 	override generateFile(String fileName, String outputCfgName, CharSequence contents) {
 		val uri = getURI(fileName, outputCfgName)
-		generateTrace(fileName, outputCfgName, contents)
 		val encoding = getEncoding(uri)
```

Comment:
Giving this another though, I think your right - the generating language should control the encoding of the files that it writes. An XMLEncodingProvider won't be able to determine an encoding for a file that was not yet written to disk anyway.
Since there are a few langs with a fixed encoding (or explicit encoding in the header), I think an overloaded IFileSystemAccess2.generateFile(.., CharSequence) may be the best we could do. It should be possible to pass an explicit Charset whenever a text is written to disk. By default the encoding provider of the generating language should be used to determine the default Charset.